### DISCUSSION TITLE
LDM Notes for July 12th, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7334] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-07-12.md

## Agenda

- Collection Literals
    - `Create` methods
    - Extension methods
- Interceptors

**Comment by mungojam** [comment]
have I understood correctly from the point on collection literals extension methods that for now the below is true?

we won't be able to do:

```c#
var a = [1, 2, 3].AsImmutableArray();
```

but we will be able to do:

```c#
var a = AsImmutableArray([1, 2, 3])
```

assuming we have a function like:

```c#
public static ImmutableArray<T> AsImmutableArray<T>(this ImmutableArray<T> array) {
     return array;
}
```

**Reply by CyrusNajmabadi** [reply]
@mungojam correct.  and that method can be a static method you bring in with a static-using.

**Comment by huoyaoyuan** [comment]
> now, instead of an out `Span` that the compiler would write into, the compiler passes in a `ReadOnlySpan`, which the collection will then copy out of

Will non-copy initialization of ImmutableArray still be supported in this release?

**Reply by CyrusNajmabadi** [reply]
Yes.  That is the intent.

**Reply by huoyaoyuan** [reply]
What API will be used then?

**Reply by CyrusNajmabadi** [reply]
https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.immutablecollectionsmarshal.asimmutablearray?view=net-8.0

**Reply by huoyaoyuan** [reply]
Is there any attribute connecting the method? Or is it a specially known initializer?
I do want to see less things specially known.

**Reply by CyrusNajmabadi** [reply]
It's specially known.
